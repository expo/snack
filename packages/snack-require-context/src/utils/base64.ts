import { Buffer } from 'buffer';

/** Convert any string to a URL-safe base64 string. */
export function encodeBase64(value: string): string {
  return Buffer.from(value)
    .toString('base64')
    .replace(/\+/g, '-')
    .replace(/\//g, '_')
    .replace(/=+$/, '');
}

/** Convert a URL-safe base64 string to its original value. */
export function decodeBase64(value: string): string {
  const base64 = value.replace(/-/g, '+').replace(/_/g, '/');
  return Buffer.from(base64, 'base64').toString('utf-8');
}
