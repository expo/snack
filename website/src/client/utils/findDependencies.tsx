import * as babel from '@babel/parser';
import { parse, types } from 'recast';
import validate from 'validate-npm-package-name';

import getFileLanguage from './getFileLanguage';

export type FileDependency = {
  isPackage: boolean;
  version?: string;
  location?: {
    fileName: string;
    startLineNumber: number;
    endLineNumber: number;
    startColumn: number;
    endColumn: number;
  };
};

export type FileDependencies = { [name: string]: FileDependency };

const parserPlugins: babel.ParserPlugin[] = [
  'asyncGenerators',
  'bigInt',
  'classProperties',
  'classPrivateProperties',
  'classPrivateMethods',
  ['decorators', { decoratorsBeforeExport: true }],
  'doExpressions',
  'dynamicImport',
  'exportDefaultFrom',
  'exportNamespaceFrom',
  'functionBind',
  'functionSent',
  'importMeta',
  'logicalAssignment',
  'nullishCoalescingOperator',
  'numericSeparator',
  'objectRestSpread',
  'optionalCatchBinding',
  'optionalChaining',
  ['pipelineOperator', { proposal: 'minimal' }],
  'throwExpressions',
];

const getVersionFromComments = (comments: { value: string }[]) => {
  return comments?.[0] && /^\s*((\d+\.)?(\d+\.)?(\*|\d+))|(LATEST)\s*$/.test(comments[0].value)
    ? comments[0].value.trim()
    : null;
};

const isValidBundle = (name: any) => {
  if (typeof name !== 'string') {
    return false;
  }

  const match = /^(?:@([^/?]+)\/)?([^@/?]+)(?:\/([^@]+))?/.exec(name);
  const fullName = match ? (match[1] ? `@${match[1]}/` : '') + match[2] : null;

  return fullName ? validate(fullName).validForOldPackages : false;
};

const removeCommentFromPath = (path: any) => {
  let node;

  if (path.node.type === 'CallExpression') {
    const { parentPath } = path;

    if (parentPath.node.type === 'VariableDeclarator') {
      node = parentPath.parentPath.node;
    } else {
      node = parentPath.node;
    }
  } else {
    node = path.node;
  }

  node.comments = node.comments || [];
  node.comments = node.comments.filter(
    (comment: any) => !(comment.type === 'CommentLine' && comment.trailing)
  );
};

const findDependencies = (
  code: string,
  filename: string,
  removeVersionComments: boolean = false
): FileDependencies => {
  const babelPlugins = [...parserPlugins];
  const language = getFileLanguage(filename);

  if (language === 'typescript') {
    babelPlugins.push('typescript');

    if (filename.endsWith('.tsx')) {
      babelPlugins.push('jsx');
    }
  } else {
    babelPlugins.push('flow', 'jsx');
  }

  const parser = {
    parse: (code: string) =>
      babel.parse(code, {
        sourceType: 'module',
        sourceFilename: filename,
        allowImportExportEverywhere: true,
        allowReturnOutsideFunction: true,
        tokens: true,
        plugins: babelPlugins,
      }),
  };

  const dependencies: FileDependencies = {};
  const ast = parse(code, { parser });

  const findModuleFromRequire = (path: any) => {
    const { callee, arguments: args } = path.node;

    let name;

    if (callee.name === 'require' && args.length === 1) {
      if (args[0].type === 'Literal' || args[0].type === 'StringLiteral') {
        name = args[0].value;
      } else if (args[0].type === 'TemplateLiteral' && args[0].quasis.length === 1) {
        name = args[0].quasis[0].value.cooked;
      }
    }

    if (name) {
      const version = getVersionFromComments(
        args[0]?.trailingComments || path.parentPath.parentPath.node.trailingComments
      );

      if (removeVersionComments) {
        removeCommentFromPath(path);
      }

      const loc = path.node.source?.loc || path.node.loc;
      dependencies[name] = {
        isPackage: isValidBundle(name),
        version: version ?? undefined,
        location: loc
          ? {
              fileName: filename,
              startLineNumber: loc.start.line,
              startColumn: loc.start.column + 1,
              endLineNumber: loc.end.line,
              endColumn: loc.end.column + 1,
            }
          : undefined,
      };
    }
  };

  const findModuleFromImport = (path: any) => {
    const name = path.node.source?.value;

    if (name) {
      const version = getVersionFromComments(path.node.trailingComments);

      if (removeVersionComments) {
        removeCommentFromPath(path);
      }

      dependencies[name] = {
        isPackage: isValidBundle(name),
        version: version ?? undefined,
        location: path.node.source.loc
          ? {
              fileName: filename,
              startLineNumber: path.node.source.loc.start.line,
              startColumn: path.node.source.loc.start.column + 1,
              endLineNumber: path.node.source.loc.end.line,
              endColumn: path.node.source.loc.end.column + 1,
            }
          : undefined,
      };
    }
  };

  try {
    types.visit(ast, {
      visitImportDeclaration(path) {
        findModuleFromImport(path);
        this.traverse(path);
      },

      visitExportNamedDeclaration(path) {
        findModuleFromImport(path);
        this.traverse(path);
      },

      visitExportAllDeclaration(path) {
        findModuleFromImport(path);
        this.traverse(path);
      },

      visitCallExpression(path) {
        findModuleFromRequire(path);
        this.traverse(path);
      },
    });
  } catch (err) {
    // Ignore error: did not recognize object of type "..."
  }

  return dependencies;
};

export default findDependencies;
