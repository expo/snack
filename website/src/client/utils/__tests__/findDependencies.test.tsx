/* eslint-env jest */

import findDependencies from '../findDependencies';

it('finds all imported modules', () => {
  const code = `
    import base64 from 'base64';
    import debounce from 'lodash/debounce';
    import { connect } from 'react-redux';
  `;

  const dependencies = findDependencies(code, 'index.js');

  expect(Object.keys(dependencies)).toEqual(['base64', 'lodash/debounce', 'react-redux']);
});

it('finds all required modules', () => {
  const code = `
    const base64 = require('base64');
    const debounce = require('lodash/debounce');
    const { connect } = require('react-redux');
  `;

  const dependencies = findDependencies(code, 'index.js');

  expect(Object.keys(dependencies)).toEqual(['base64', 'lodash/debounce', 'react-redux']);
});

it('finds all required modules with backticks', () => {
  const code = `
    const base64 = require(\`base64\`);
    const debounce = require(\`lodash/debounce\`);
    const { connect } = require(\`react-redux\`);
  `;

  const dependencies = findDependencies(code, 'index.js');

  expect(Object.keys(dependencies)).toEqual(['base64', 'lodash/debounce', 'react-redux']);
});

it('finds all required modules with version comments', () => {
  const code = `
    const base64 = require('base64'); // 2.4.1
    const { connect } = require(\`react-redux\`); // 3.5.2
  `;

  const dependencies = findDependencies(code, 'index.js');

  expect(dependencies).toMatchSnapshot();
});

it('finds dependencies using all import styles', () => {
  const code = `
    import v from "mod1"
    import * as ns from "mod2";
    import {x} from "mod3";
    import {x as v2} from "mod4";
    import "mod5";

    export {x} from "mod6";
    export {x as v} from "mod7";
    export * from "mod8";

    export default 7;
    export const value = 6;
    const otherValue = 5;
    export { otherValue }
  `;

  const dependencies = findDependencies(code, 'index.js');

  expect(Object.keys(dependencies)).toEqual([
    'mod1',
    'mod2',
    'mod3',
    'mod4',
    'mod5',
    'mod6',
    'mod7',
    'mod8',
  ]);
});

it('finds dependencies using all import styles in TypeScript file', () => {
  const code = `
    import v from "mod1"
    import * as ns from "mod2";
    import {x} from "mod3";
    import {x as v2} from "mod4";
    import "mod5";

    export {x} from "mod6";
    export {x as v} from "mod7";
    export * from "mod8";

    export default 7;
    export const value = 6;
    const otherValue = 5;
    export { otherValue };

    // TypeScript features
    const assets: number[] = [];
  `;

  const dependencies = findDependencies(code, 'index.tsx');

  expect(Object.keys(dependencies)).toEqual([
    'mod1',
    'mod2',
    'mod3',
    'mod4',
    'mod5',
    'mod6',
    'mod7',
    'mod8',
  ]);
});

it('finds dependencies with version comments', () => {
  const code = `
    import v from "mod1" // 2.4.1
    import * as ns from "mod2"; // 3.5.2
    import {x} from "mod3"; // 4.6.3
    import {x as v2} from "mod4"; // 5.7.4
    import "mod5"; // 6.8.5

    export {x} from "mod6"; // 7.9.6
    export {x as v} from "mod7"; // 8.0.7
    export * from "mod8"; // 9.1.8
  `;

  const dependencies = findDependencies(code, 'index.js');

  expect(dependencies).toMatchSnapshot();
});

it("doesn't parse non-static and invalid requires", () => {
  const code = `
    const base64 = require();
    const uuid = require('');
    const debounce = require('debounce', null);
    const moment = require(\`\${name}\`);
    const leftpad = require(\`left-\${name}\`);
    const core = require('/core');
    const promise = require('promise\\nme');
    const bluebird = require(\`blue
      bird
    \`);
    const ten = require(10);
  `;

  const dependencies = findDependencies(code, 'index.js');

  expect(Object.values(dependencies).filter((dep) => dep.isPackage)).toEqual([]);
});

it('finds dependencies in JSX files', () => {
  const code = `
    import * as React from "react";
    import { View, Text } from "react-native";
    import v from "mod1" // 2.4.1
    import * as ns from "mod2"; // 3.5.2
    import {x} from "mod3"; // 4.6.3
    import {x as v2} from "mod4"; // 5.7.4
    import "mod5"; // 6.8.5
    
    export default function App() {
      return (
        <View>
          <Text>
            Change code in the editor and watch it change on your phone! Save to get a shareable url.
          </Text>
        </View>
      );
    }
  `;

  const dependencies = findDependencies(code, 'index.js');

  expect(dependencies).toMatchSnapshot();
});
