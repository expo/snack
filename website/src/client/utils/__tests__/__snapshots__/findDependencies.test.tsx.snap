// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`finds all required modules with version comments 1`] = `
Object {
  "base64": Object {
    "isPackage": true,
    "location": Object {
      "endColumn": 37,
      "endLineNumber": 2,
      "fileName": "index.js",
      "startColumn": 20,
      "startLineNumber": 2,
    },
    "version": "2.4.1",
  },
  "react-redux": Object {
    "isPackage": true,
    "location": Object {
      "endColumn": 47,
      "endLineNumber": 3,
      "fileName": "index.js",
      "startColumn": 25,
      "startLineNumber": 3,
    },
    "version": "3.5.2",
  },
}
`;

exports[`finds dependencies in JSX files 1`] = `
Object {
  "mod1": Object {
    "isPackage": true,
    "location": Object {
      "endColumn": 25,
      "endLineNumber": 4,
      "fileName": "index.js",
      "startColumn": 19,
      "startLineNumber": 4,
    },
    "version": "2.4.1",
  },
  "mod2": Object {
    "isPackage": true,
    "location": Object {
      "endColumn": 31,
      "endLineNumber": 5,
      "fileName": "index.js",
      "startColumn": 25,
      "startLineNumber": 5,
    },
    "version": "3.5.2",
  },
  "mod3": Object {
    "isPackage": true,
    "location": Object {
      "endColumn": 27,
      "endLineNumber": 6,
      "fileName": "index.js",
      "startColumn": 21,
      "startLineNumber": 6,
    },
    "version": "4.6.3",
  },
  "mod4": Object {
    "isPackage": true,
    "location": Object {
      "endColumn": 33,
      "endLineNumber": 7,
      "fileName": "index.js",
      "startColumn": 27,
      "startLineNumber": 7,
    },
    "version": "5.7.4",
  },
  "mod5": Object {
    "isPackage": true,
    "location": Object {
      "endColumn": 18,
      "endLineNumber": 8,
      "fileName": "index.js",
      "startColumn": 12,
      "startLineNumber": 8,
    },
    "version": "6.8.5",
  },
  "react": Object {
    "isPackage": true,
    "location": Object {
      "endColumn": 35,
      "endLineNumber": 2,
      "fileName": "index.js",
      "startColumn": 28,
      "startLineNumber": 2,
    },
    "version": undefined,
  },
  "react-native": Object {
    "isPackage": true,
    "location": Object {
      "endColumn": 46,
      "endLineNumber": 3,
      "fileName": "index.js",
      "startColumn": 32,
      "startLineNumber": 3,
    },
    "version": undefined,
  },
}
`;

exports[`finds dependencies with version comments 1`] = `
Object {
  "mod1": Object {
    "isPackage": true,
    "location": Object {
      "endColumn": 25,
      "endLineNumber": 2,
      "fileName": "index.js",
      "startColumn": 19,
      "startLineNumber": 2,
    },
    "version": "2.4.1",
  },
  "mod2": Object {
    "isPackage": true,
    "location": Object {
      "endColumn": 31,
      "endLineNumber": 3,
      "fileName": "index.js",
      "startColumn": 25,
      "startLineNumber": 3,
    },
    "version": "3.5.2",
  },
  "mod3": Object {
    "isPackage": true,
    "location": Object {
      "endColumn": 27,
      "endLineNumber": 4,
      "fileName": "index.js",
      "startColumn": 21,
      "startLineNumber": 4,
    },
    "version": "4.6.3",
  },
  "mod4": Object {
    "isPackage": true,
    "location": Object {
      "endColumn": 33,
      "endLineNumber": 5,
      "fileName": "index.js",
      "startColumn": 27,
      "startLineNumber": 5,
    },
    "version": "5.7.4",
  },
  "mod5": Object {
    "isPackage": true,
    "location": Object {
      "endColumn": 18,
      "endLineNumber": 6,
      "fileName": "index.js",
      "startColumn": 12,
      "startLineNumber": 6,
    },
    "version": "6.8.5",
  },
  "mod6": Object {
    "isPackage": true,
    "location": Object {
      "endColumn": 27,
      "endLineNumber": 8,
      "fileName": "index.js",
      "startColumn": 21,
      "startLineNumber": 8,
    },
    "version": "7.9.6",
  },
  "mod7": Object {
    "isPackage": true,
    "location": Object {
      "endColumn": 32,
      "endLineNumber": 9,
      "fileName": "index.js",
      "startColumn": 26,
      "startLineNumber": 9,
    },
    "version": "8.0.7",
  },
  "mod8": Object {
    "isPackage": true,
    "location": Object {
      "endColumn": 25,
      "endLineNumber": 10,
      "fileName": "index.js",
      "startColumn": 19,
      "startLineNumber": 10,
    },
    "version": "9.1.8",
  },
}
`;
